haskell/private/haskell_impl.bzl

```diff
@@ -190,6 +184,13 @@ def _haskell_binary_common_impl(ctx, is_test):
     # Make shell tools available.
     posix = ctx.toolchains["@rules_sh//sh/posix:toolchain_type"]
 
+    # Determine file directories.
+    interfaces_dir = paths.join("_iface", hs.name)
+    objects_dir = paths.join("_obj", hs.name)
+
+    module_interfaces, module_objects = build_haskell_modules(ctx, hs, cc, posix, "", interfaces_dir, objects_dir)
+    extra_objects = module_objects.to_list()
```

Comment:
Not for this PR, but, I think we can avoid the `.to_list()` here and keep this as a `depset`. As far as I see one use-case is to add this to some action inputs, which can easily be done with a `depset` as well, and the other is [constructing the `objects_dir_manifest`](https://github.com/tweag/rules_haskell/blob/9efbbbecbc37c7c2fb72058f018da90a16621417/haskell/private/actions/link.bzl#L68) where we could use the `depset` directly by passing it to an args object instead of flattening it and splicing it into the command string.

Reply:
I explored this, the calls in `actions/link.bzl` filter the objects files by extension to link statically or dynamically. Doing the filtering without expanding probably will require putting `.o` and `.dyn_o` in different depsets to begin with.

Reply:
Probably a good time to revise this is when adding TH support.

Reply:
> Doing the filtering without expanding probably will require putting `.o` and `.dyn_o` in different depsets to begin with.

Ah, yes, that sounds like a reasonable approach.

> Probably a good time to revise this is when adding TH support.

Agreed.